File: src/sp/it/pl/util/functional/StrExF.java
package sp.it.pl.util.functional;

import net.objecthunter.exp4j.Expression;
import net.objecthunter.exp4j.ExpressionBuilder;
import net.objecthunter.exp4j.ValidationResult;
import sp.it.pl.util.functional.Functors.Ƒ1;
import sp.it.pl.util.parsing.ParsesFromString;
import sp.it.pl.util.parsing.ParsesToString;
import sp.it.pl.util.parsing.StringParseStrategy;
import sp.it.pl.util.parsing.StringParseStrategy.From;
import sp.it.pl.util.parsing.StringParseStrategy.To;

/**
 * String expression function
 */
@StringParseStrategy(
		from = From.ANNOTATED_METHOD, to = To.TO_STRING_METHOD,
		exFrom = {IllegalStateException.class, IllegalArgumentException.class}
)
public class StrExF implements Ƒ1<Double,Double> {
	private final String ex;
	private final Expression e;

	@ParsesFromString
	public StrExF(String s) {
		try {
			ex = s;
			e = new ExpressionBuilder(s).variables("x").build();
			ValidationResult v = e.validate(false);
			if (!v.isValid())
				throw new Exception(v.getErrors().get(0));
		} catch (Exception e) {
			throw new IllegalStateException(e);
		}
	}

	@Override
	public Double apply(Double queryParam) {
		return e.setVariable("x", queryParam).evaluate();
	}

	@ParsesToString
	@Override
	public String toString() {
		return ex;
	}

}

File: src/sp/it/pl/util/math/StrExF.kt
package sp.it.pl.util.math

import net.objecthunter.exp4j.Expression
import net.objecthunter.exp4j.ExpressionBuilder
import sp.it.pl.util.dev.Dependency
import sp.it.pl.util.functional.Functors.Ƒ1
import sp.it.pl.util.functional.Try

/** Function created from string expression. */
class StrExF(s: String): Ƒ1<Double, Double> {
    private val expression: String
    private val e: Expression

    init {
        try {
            expression = s
            e = ExpressionBuilder(s).variables("x").build()
            val v = e.validate(false)
            if (!v.isValid)
                throw Exception(v.errors[0])
        } catch (e: Exception) {
            throw IllegalStateException(e)
        }
    }

    override fun apply(queryParam: Double): Double = e.setVariable("x", queryParam).evaluate()

    @Dependency("fromString")
    override fun toString() = expression

    companion object {
        @Dependency("toString")
        fun fromString(s: String): Try<StrExF, Exception> {
            return try {
                Try.ok(StrExF(s))
            } catch (e: Exception) {
                Try.error(e)
            }
        }
    }

}